### PR TITLE
Internal documentation

### PR DESCRIPTION
**Pull request by sitano**
What do you think? Is it fine? I will write a few sections more.

### CODE REVIEW

File: Documentation/internal/README.md [hunk 1]
```diff
@@ -1,5 +1,160 @@
...
+
+`delve` consists of the debugger engine which can read and understand Go process memory (located
+in package `proc`), [dwarf](http://dwarfstd.org/doc/dwarf-2.0.0.pdf) 
+[parser](https://golang.org/pkg/debug/dwarf/) and wrapper (go specifics), `terminal` helper,
+`service` set of the RPC servers with user friendly `service/debugger` API.
+
+# `proc` package
+
+The `proc` package is the heart of the `delve` program. It implements low-level debugging interface,
+remote process and memory manipulation, disasm and etc. `proc` is based on the `ptrace(2)` syscall
```
Comment (aarzilli) [inline Documentation/internal/README.md]: `proc` uses ptrace on linux, ptrace + mach specific syscalls on OS X and WaitForDebugEvent+ContinueDebugEvent +SuspendThread+ResumeThread on Windows.

File: Documentation/internal/README.md [hunk 2]
```diff
@@ -1,5 +1,160 @@
...
+* [ptrace - WikiPedia](https://en.wikipedia.org/wiki/Ptrace)
+
+## `ptrace` binding
+
+`ptrace` implementation uses 1 separate goroutine per debugging process (instance of the `ptrace`). To
+guarantee safe `ptrace` environment `runtime.LockOSThread` is used. Commands received via Go channel.
+
+# `dwarf` 
+
+`dwarf` stands for DWARF Debugging Information Format. The 6l and 8l linkers emit DWARF3 debugging information
```
Comment (aarzilli) [inline Documentation/internal/README.md]: 6l and 8l don't exist anymore, I'm not even sure we support any version of go where 6l and 8l still exist. The linker is just called `link` now

File: Documentation/internal/README.md [hunk 3]
```diff
@@ -1,5 +1,160 @@
...
+
+- load a Go program in GDB version 7.x,
+- list all Go, C, and assembly source files by line (parts of the Go runtime are written in C and assembly),
+- set breakpoints by line and step through the code,
+- print stack traces and inspect stack frames, and
+- find the addresses and print the contents of most variables.
+
+Link:
+
+* [DWARF Debugging Information Format](http://dwarfstd.org/doc/dwarf-2.0.0.pdf)
```
Comment (aarzilli) [inline Documentation/internal/README.md]: This should link to [DWARFv3](http://dwarfstd.org/doc/Dwarf3.pdf)

File: Documentation/internal/README.md [hunk 4]
```diff
@@ -1,5 +1,160 @@
...
+- `ptrace` [syscall](http://tldp.org/LDP/LG/issue81/sandeep.html) go binding used to do low level 
+  control of the remote processes, memory IO, start/stop/stepping execution and etc.
+- `Registers` interface
+- `Stack` iterator over stack frames
+- `Thread` management
+- `Types` parser
+- `Variables` parser
+
+Cross platform debug support is implemented for Linux, Darwin (MacOSX) and Windows. 
+Windows lacks support of process attach for now.
```
Comment (alexbrainman) [inline Documentation/internal/README.md]: You should be able to attach to a running Windows process if you use current tip. See f09ef23f9ec650f18f2530d7164229ea4a119700 for details.

**Comment by sitano** [comment]
@alexbrainman you think its worth having more high-level overview without stitching it with a code refs?

**Comment by alexbrainman** [comment]
> @alexbrainman you think its worth having more high-level overview without stitching it with a code refs?

I do not know what the purpose of Documentation/internal/README.md is. I will let authors answer your question. I just think that Documentation/internal/README.md might get out of sync with the source over time.

Alex

**Comment by derekparker** [comment]
@sitano I appreciate this type of documentation, but have the same concerns as @alexbrainman regarding this information getting out of sync over time.

What I think may be in the same vein, but less likely to diverge, is a higher level document explaining some of the core concepts and where to learn more about them. That could significantly reduce the barrier to entry for new contributors, and we could feature that doc in the main README, point it out during talks, etc.

Thanks for taking the time to dig in and construct this!

**Comment by sitano** [comment]
Nice overview https://backtrace.io/blog/blog/2016/11/23/building-a-go-debugger/

**Pull request closed without merging**